File: loaders/typescript/SkillLoader.ts
/**
 * Universal Skill Loader - TypeScript Implementation
 *
 * Load SKILL.md files and convert them to various platform formats.
 */

import { readFile } from 'fs/promises';
import frontmatter from 'front-matter';
import { glob } from 'glob';

export interface SkillConfig {
  name: string;
  description: string;
  version: string;
  author: string;
  triggers: string[];
  tools: string[];
  content: string;
  executionSteps: string[];
}

export interface AgentSDKConfig {
  name: string;
  description: string;
  instructions: string;
  tools: string[];
}

export interface OpenClawConfig {
  name: string;
  description: string;
  version: string;
  skills: OpenClawSkill[];
}

export interface OpenClawSkill {
  name: string;
  description: string;
  tools: { name: string }[];
}

export class SkillLoader {
  private toolMap: Map<string, string>;

  constructor() {
    this.toolMap = this.buildToolMap();
  }

  /**
   * Load a SKILL.md file
   */
  async load(skillPath: string): Promise<SkillConfig> {
    const content = await readFile(skillPath, 'utf-8');
    const { attributes, body } = frontmatter(content);

    const metadata = attributes as Record<string, any>;

    return {
      name: metadata.name || 'unknown',
      description: metadata.description || '',
      version: metadata.version || '1.0.0',
      author: metadata.author || '',
      triggers: metadata.triggers || [],
      tools: metadata.tools || [],
      content: body,
      executionSteps: this.parseExecutionSteps(body)
    };
  }

  /**
   * Convert to Agent SDK format
   */
  toAgentSDK(config: SkillConfig): AgentSDKConfig {
    return {
      name: config.name,
      description: config.description,
      instructions: this.generateInstructions(config),
      tools: this.mapTools(config.tools)
    };
  }

  /**
   * Convert to OpenClaw format
   */
  toOpenClaw(config: SkillConfig): OpenClawConfig {
    return {
      name: config.name,
      description: config.description,
      version: config.version,
      skills: [{
        name: 'execute',
        description: config.description,
        tools: config.tools.map(tool => ({
          name: this.toolMap.get(tool, tool)
        }))
      }]
    };
  }

  /**
   * Parse execution steps from markdown content
   */
  private parseExecutionSteps(content: string): string[] {
    const steps: string[] = [];
    const lines = content.split('\n');
    let inExecutionSection = false;

    for (const line of lines) {
      const trimmed = line.trim();

      if (trimmed.startsWith('## Execution Steps') || trimmed.startsWith('## 執行步驟')) {
        inExecutionSection = true;
        continue;
      }

      if (inExecutionSection && trimmed.startsWith('#')) {
        break;
      }

      if (inExecutionSection && trimmed) {
        steps.push(trimmed);
      }
    }

    return steps;
  }

  /**
   * Generate system prompt from SkillConfig
   */
  private generateInstructions(config: SkillConfig): string {
    const steps = config.executionSteps
      .map((step, i) => `${i + 1}. ${step}`)
      .join('\n');

    return `You are ${config.name}

${config.description}

## Available Tools
${config.tools.join(', ')}

## Execution Steps
${steps}

## Additional Context
${config.content}`;
  }

  /**
   * Map tool names to target platform format
   */
  private mapTools(tools: string[]): string[] {
    return tools.map(tool => this.toolMap.get(tool, tool));
  }

  /**
   * Build tool name mapping
   */
  private buildToolMap(): Map<string, string> {
    return new Map([
      // Claude Code → Agent SDK
      ['WebSearch', 'web_search'],
      ['WebFetch', 'web_fetch'],
      ['Read', 'read_file'],
      ['Write', 'write_file'],
      ['Grep', 'grep'],
      ['Bash', 'bash_command'],

      // Generic mappings
      ['web-search', 'web_search'],
      ['web-fetch', 'web_fetch'],
    ]);
  }

  /**
   * Load all Skills from a directory
   */
  async loadFromDirectory(dir: string): Promise<SkillConfig[]> {
    const files = await glob(`${dir}/**/SKILL.md`);
    const configs: SkillConfig[] = [];

    for (const file of files) {
      const config = await this.load(file);
      configs.push(config);
    }

    return configs;
  }
}

/**
 * CLI interface
 */
async function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log('Usage: npx ts-node skill_loader.ts <skill-path> [format]');
    console.log('Formats: agent-sdk, openclaw');
    process.exit(1);
  }

  const [skillPath, format = 'agent-sdk'] = args;

  const loader = new SkillLoader();
  const config = await loader.load(skillPath);

  let result;

  if (format === 'agent-sdk') {
    result = loader.toAgentSDK(config);
  } else if (format === 'openclaw') {
    result = loader.toOpenClaw(config);
  } else {
    console.error(`Unknown format: ${format}`);
    process.exit(1);
  }

  console.log(JSON.stringify(result, null, 2));
}

if (require.main === module) {
  main().catch(console.error);
}

export { SkillLoader };

File: loaders/python/skill_loader.py
"""
Universal Skill Loader - Python Implementation
Load SKILL.md files and convert them to Agent SDK agents.
"""

import os
import json
import frontmatter
from typing import Dict, List, Any, Optional


class SkillConfig:
    """Parsed SKILL.md configuration"""
    def __init__(self, metadata: Dict[str, Any], content: str):
        self.name = metadata.get('name', 'unknown')
        self.description = metadata.get('description', '')
        self.version = metadata.get('version', '1.0.0')
        self.author = metadata.get('author', '')
        self.triggers = metadata.get('triggers', [])
        self.tools = metadata.get('tools', [])
        self.content = content


class SkillLoader:
    """Load SKILL.md files and convert to various formats"""
    def __init__(self):
        self.tool_map = {
            'WebSearch': 'web_search',
            'WebFetch': 'web_fetch',
            'Read': 'read_file',
            'Write': 'write_file',
            'Grep': 'grep',
            'Bash': 'bash_command',
        }

    def load(self, skill_path: str) -> SkillConfig:
        """Load a SKILL.md file"""
        with open(skill_path, 'r', encoding='utf-8') as f:
            content = f.read()
        post = frontmatter.loads(content)
        return SkillConfig(post.metadata, post.content)

    def to_agent_sdk(self, config: SkillConfig) -> Dict[str, Any]:
        """Convert to Agent SDK format"""
        return {
            'name': config.name,
            'description': config.description,
            'instructions': f"You are {config.name}\n\n{config.description}",
            'tools': [self.tool_map.get(t, t) for t in config.tools]
        }

    def to_openclaw(self, config: SkillConfig) -> Dict[str, Any]:
        """Convert to OpenClaw format"""
        return {
            'name': config.name,
            'description': config.description,
            'version': config.version,
            'skills': [{
                'name': 'execute',
                'tools': [{'name': self.tool_map.get(t, t)} for t in config.tools]
            }]
        }


def main():
    import sys
    if len(sys.argv) < 2:
        print("Usage: python skill_loader.py <skill-path> [format]")
        sys.exit(1)

    loader = SkillLoader()
    config = loader.load(sys.argv[1])
    format = sys.argv[2] if len(sys.argv) > 2 else 'agent-sdk'

    if format == 'agent-sdk':
        result = loader.to_agent_sdk(config)
    elif format == 'openclaw':
        result = loader.to_openclaw(config)
    else:
        print(f"Unknown format: {format}")
        sys.exit(1)

    print(json.dumps(result, indent=2))


if __name__ == '__main__':
    main()
